File: users-service/src/main/java/ru/dsis/atms/users/repository/RolesRepository.java
package ru.dsis.atms.users.repository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import ru.dsis.atms.users.dao.data.RoleData;
import ru.dsis.atms.users.dao.postgres.PostgresRole;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

@Repository
public class RolesRepository {
    @Autowired
    private JdbcTemplate jdbcTemplate;

    public void save(RoleData role) {
        String sql = "INSERT INTO roles (role_name) VALUES (?)";
        jdbcTemplate.update(sql, role.getRoleName());
    }

    public List<PostgresRole> findAll() {
        String sql = "SELECT * FROM roles";
        return jdbcTemplate.query(sql, new RoleRowMapper());
    }
//
//    public List<PostgresRole> findRoleByUsername(String username) {
//        String sql = "SELECT r FROM roles r " +
//                     "JOIN user_roles ur ON r.id = ur.role_id " +
//                     "JOIN users u ON u.id = ur.user_id " +
//                     "WHERE u.username = ?";
//        return jdbcTemplate.query(sql, new RoleRowMapper(), username);
//    }

    public PostgresRole findByRoleName(String roleName) {
        String sql = "SELECT * FROM roles WHERE role_name = ?";
        return jdbcTemplate.queryForObject(sql, new RoleRowMapper(), roleName);
    }

    private static class RoleRowMapper implements RowMapper<PostgresRole> {
        @Override
        public PostgresRole mapRow(ResultSet rs, int rowNum) throws SQLException {
            PostgresRole role = new PostgresRole();
            role.setId(rs.getLong("id"));
            role.setRoleName(rs.getString("role_name"));
            return role;
        }
    }
}

File: users-service/src/main/kotlin/ru/dsis/atms/users/repository/RolesRepository.kt
package ru.dsis.atms.users.repository

import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import org.springframework.stereotype.Repository
import ru.dsis.atms.users.dao.postgres.PostgresRole
import java.sql.ResultSet
import java.sql.SQLException

@Repository
class RolesRepository(val jdbcTemplate: JdbcTemplate) {
    fun findAll(): List<PostgresRole> {
        val sql = """
            SELECT r.id as role_id, r.role_name, r.role_description FROM roles r
        """.trimIndent()
        return jdbcTemplate.query(sql, RoleRowMapper())
    }

    fun findByRoleId(roleId: Long): PostgresRole? {
        val sql = """
            SELECT r.id as role_id, r.role_name, r.role_description FROM roles r
            WHERE r.id = ?
        """.trimIndent()
        return jdbcTemplate.queryForObject(sql, RoleRowMapper(), roleId)
    }

    fun findByRoleName(roleName: String): PostgresRole? {
        val sql = """
            SELECT r.id as role_id, r.role_name, r.role_description FROM roles r
            WHERE r.role_name = ?
        """.trimIndent()
        return jdbcTemplate.queryForObject(sql, RoleRowMapper(), roleName)
    }

    fun findByUserId(userId: Long): PostgresRole? {
        val sql = """
            SELECT r.id as role_id, r.role_name, r.role_description FROM roles r
            INNER JOIN users u ON r.role_name = u.role_name
            WHERE u.id = ?
        """.trimIndent()
        return jdbcTemplate.queryForObject(sql, RoleRowMapper(), userId)
    }

    fun findByUsername(username: String): PostgresRole? {
        val sql = """
            SELECT r.id as role_id, r.role_name, r.role_description FROM roles r
            INNER JOIN users u ON r.role_name = u.role_name
            WHERE u.username = ?
        """.trimIndent()
        return jdbcTemplate.queryForObject(sql, RoleRowMapper(), username)
    }

    private class RoleRowMapper : RowMapper<PostgresRole> {
        @Throws(SQLException::class)
        override fun mapRow(rs: ResultSet, rowNum: Int): PostgresRole {
            val role = PostgresRole()
            role.id = rs.getLong("role_id")
            role.roleName = rs.getString("role_name")
            role.description = rs.getString("role_description")
            return role
        }
    }
}